File: cpp/meetup/meetup.cpp
#include "meetup.h"

namespace meetup {

scheduler::scheduler(boost::gregorian::greg_month month, int year)
    : month_(month), year_(year) {}

// Helper method to find a "teenth" day (13th-19th) that falls on a specific weekday
boost::gregorian::date scheduler::find_teenth(boost::gregorian::greg_weekday weekday) const {
    // Start from the 13th of the month
    boost::gregorian::date start_date(year_, month_, 13);
    
    // Calculate days to add to get to the desired weekday
    int days_to_add = (weekday.as_number() - start_date.day_of_week().as_number() + 7) % 7;
    
    // Return the date of the desired weekday
    return start_date + boost::gregorian::days(days_to_add);
}

// Helper method to find the nth occurrence of a weekday in the month
boost::gregorian::date scheduler::find_nth_weekday(int n, boost::gregorian::greg_weekday weekday) const {
    // Start from the first day of the month
    boost::gregorian::date start_date(year_, month_, 1);
    
    // Calculate days to add to get to the first occurrence of the desired weekday
    int days_to_first = (weekday.as_number() - start_date.day_of_week().as_number() + 7) % 7;
    
    // Calculate the date of the nth occurrence
    return start_date + boost::gregorian::days(days_to_first + (n - 1) * 7);
}

// Helper method to find the last occurrence of a weekday in the month
boost::gregorian::date scheduler::find_last_weekday(boost::gregorian::greg_weekday weekday) const {
    // Get the last day of the month
    boost::gregorian::date end_of_month = boost::gregorian::date(year_, month_, 1).end_of_month();
    
    // Calculate days to subtract to get to the last occurrence of the desired weekday
    int days_to_subtract = (end_of_month.day_of_week().as_number() - weekday.as_number() + 7) % 7;
    
    // Return the date of the last occurrence of the desired weekday
    return end_of_month - boost::gregorian::days(days_to_subtract);
}

// Teenth methods
boost::gregorian::date scheduler::monteenth() const {
    return find_teenth(boost::gregorian::Monday);
}

boost::gregorian::date scheduler::tuesteenth() const {
    return find_teenth(boost::gregorian::Tuesday);
}

boost::gregorian::date scheduler::wednesteenth() const {
    return find_teenth(boost::gregorian::Wednesday);
}

boost::gregorian::date scheduler::thursteenth() const {
    return find_teenth(boost::gregorian::Thursday);
}

boost::gregorian::date scheduler::friteenth() const {
    return find_teenth(boost::gregorian::Friday);
}

boost::gregorian::date scheduler::saturteenth() const {
    return find_teenth(boost::gregorian::Saturday);
}

boost::gregorian::date scheduler::sunteenth() const {
    return find_teenth(boost::gregorian::Sunday);
}

// First weekday methods
boost::gregorian::date scheduler::first_monday() const {
    return find_nth_weekday(1, boost::gregorian::Monday);
}

boost::gregorian::date scheduler::first_tuesday() const {
    return find_nth_weekday(1, boost::gregorian::Tuesday);
}

boost::gregorian::date scheduler::first_wednesday() const {
    return find_nth_weekday(1, boost::gregorian::Wednesday);
}

boost::gregorian::date scheduler::first_thursday() const {
    return find_nth_weekday(1, boost::gregorian::Thursday);
}

boost::gregorian::date scheduler::first_friday() const {
    return find_nth_weekday(1, boost::gregorian::Friday);
}

boost::gregorian::date scheduler::first_saturday() const {
    return find_nth_weekday(1, boost::gregorian::Saturday);
}

boost::gregorian::date scheduler::first_sunday() const {
    return find_nth_weekday(1, boost::gregorian::Sunday);
}

// Second weekday methods
boost::gregorian::date scheduler::second_monday() const {
    return find_nth_weekday(2, boost::gregorian::Monday);
}

boost::gregorian::date scheduler::second_tuesday() const {
    return find_nth_weekday(2, boost::gregorian::Tuesday);
}

boost::gregorian::date scheduler::second_wednesday() const {
    return find_nth_weekday(2, boost::gregorian::Wednesday);
}

boost::gregorian::date scheduler::second_thursday() const {
    return find_nth_weekday(2, boost::gregorian::Thursday);
}

boost::gregorian::date scheduler::second_friday() const {
    return find_nth_weekday(2, boost::gregorian::Friday);
}

boost::gregorian::date scheduler::second_saturday() const {
    return find_nth_weekday(2, boost::gregorian::Saturday);
}

boost::gregorian::date scheduler::second_sunday() const {
    return find_nth_weekday(2, boost::gregorian::Sunday);
}

// Third weekday methods
boost::gregorian::date scheduler::third_monday() const {
    return find_nth_weekday(3, boost::gregorian::Monday);
}

boost::gregorian::date scheduler::third_tuesday() const {
    return find_nth_weekday(3, boost::gregorian::Tuesday);
}

boost::gregorian::date scheduler::third_wednesday() const {
    return find_nth_weekday(3, boost::gregorian::Wednesday);
}

boost::gregorian::date scheduler::third_thursday() const {
    return find_nth_weekday(3, boost::gregorian::Thursday);
}

boost::gregorian::date scheduler::third_friday() const {
    return find_nth_weekday(3, boost::gregorian::Friday);
}

boost::gregorian::date scheduler::third_saturday() const {
    return find_nth_weekday(3, boost::gregorian::Saturday);
}

boost::gregorian::date scheduler::third_sunday() const {
    return find_nth_weekday(3, boost::gregorian::Sunday);
}

// Fourth weekday methods
boost::gregorian::date scheduler::fourth_monday() const {
    return find_nth_weekday(4, boost::gregorian::Monday);
}

boost::gregorian::date scheduler::fourth_tuesday() const {
    return find_nth_weekday(4, boost::gregorian::Tuesday);
}

boost::gregorian::date scheduler::fourth_wednesday() const {
    return find_nth_weekday(4, boost::gregorian::Wednesday);
}

boost::gregorian::date scheduler::fourth_thursday() const {
    return find_nth_weekday(4, boost::gregorian::Thursday);
}

boost::gregorian::date scheduler::fourth_friday() const {
    return find_nth_weekday(4, boost::gregorian::Friday);
}

boost::gregorian::date scheduler::fourth_saturday() const {
    return find_nth_weekday(4, boost::gregorian::Saturday);
}

boost::gregorian::date scheduler::fourth_sunday() const {
    return find_nth_weekday(4, boost::gregorian::Sunday);
}

// Last weekday methods
boost::gregorian::date scheduler::last_monday() const {
    return find_last_weekday(boost::gregorian::Monday);
}

boost::gregorian::date scheduler::last_tuesday() const {
    return find_last_weekday(boost::gregorian::Tuesday);
}

boost::gregorian::date scheduler::last_wednesday() const {
    return find_last_weekday(boost::gregorian::Wednesday);
}

boost::gregorian::date scheduler::last_thursday() const {
    return find_last_weekday(boost::gregorian::Thursday);
}

boost::gregorian::date scheduler::last_friday() const {
    return find_last_weekday(boost::gregorian::Friday);
}

boost::gregorian::date scheduler::last_saturday() const {
    return find_last_weekday(boost::gregorian::Saturday);
}

boost::gregorian::date scheduler::last_sunday() const {
    return find_last_weekday(boost::gregorian::Sunday);
}

}  // namespace meetup

File: javascript/meetup/meetup.js
export const meetup = (year, month, descriptor, weekday) => {
  // Map weekday names to their numeric values (0 = Sunday, 1 = Monday, etc.)
  const weekdayMap = {
    'Sunday': 0,
    'Monday': 1,
    'Tuesday': 2,
    'Wednesday': 3,
    'Thursday': 4,
    'Friday': 5,
    'Saturday': 6
  };

  // Get the numeric value for the specified weekday
  const targetWeekday = weekdayMap[weekday];

  // Handle different descriptors
  switch (descriptor) {
    case 'teenth': {
      // The 'teenth' days are from 13 to 19
      // Start from the 13th and find the first occurrence of the target weekday
      for (let day = 13; day <= 19; day++) {
        const date = new Date(year, month - 1, day);
        if (date.getDay() === targetWeekday) {
          return date;
        }
      }
      break;
    }
    
    case 'first':
    case 'second':
    case 'third':
    case 'fourth': {
      // Map descriptor to the number of occurrences to find
      const occurrenceMap = {
        'first': 1,
        'second': 2,
        'third': 3,
        'fourth': 4
      };
      
      const targetOccurrence = occurrenceMap[descriptor];
      let currentOccurrence = 0;
      
      // Start from the first day of the month and count occurrences
      for (let day = 1; day <= 31; day++) {
        const date = new Date(year, month - 1, day);
        
        // If we've gone to the next month, break
        if (date.getMonth() !== month - 1) {
          break;
        }
        
        // If this day is the target weekday, increment the occurrence counter
        if (date.getDay() === targetWeekday) {
          currentOccurrence++;
          
          // If we've found the target occurrence, return the date
          if (currentOccurrence === targetOccurrence) {
            return date;
          }
        }
      }
      break;
    }
    
    case 'last': {
      // Start from the last possible day of the month and work backwards
      // Get the first day of the next month and subtract 1 to get the last day of the current month
      const lastDay = new Date(year, month, 0).getDate();
      
      // Start from the last day and work backwards
      for (let day = lastDay; day >= 1; day--) {
        const date = new Date(year, month - 1, day);
        if (date.getDay() === targetWeekday) {
          return date;
        }
      }
      break;
    }
  }
  
  // If we get here, something went wrong
  throw new Error('Could not find the specified meetup date');
};
